src/main/java/org/generationcp/middleware/api/brapi/ObservationServiceBrapiImpl.java

```diff
@@ -0,0 +1,117 @@
+			for(final ObservationDto observationDto: observationDtos) {
+				observationDto.setExternalReferences(externalReferencesMap.get(observationDto.getObservationDbId()));
+			}
+		}
+		return observationDtos;
+	}
+
+	@Override
+	public long countObservations(final ObservationSearchRequestDto observationSearchRequestDto) {
+		return this.daoFactory.getPhenotypeDAO().countObservations(observationSearchRequestDto);
+	}
+
+	@Override
+	public List<Integer> createObservations(final List<ObservationDto> observations) {
+
+		final List<Integer> observationDbIds = new ArrayList<>();
+		final Map<String, ExperimentModel> experimentModelMap = this.daoFactory.getExperimentDao()
+			.getByObsUnitIds(observations.stream().map(ObservationDto::getObservationUnitDbId).collect(Collectors.toList())).stream()
+			.collect(Collectors.toMap(ExperimentModel::getObsUnitId, Function.identity()));
+		final Map<Integer, List<ValueReference>> validValuesForCategoricalVariables =
+			this.getCategoriesForCategoricalVariables(observations);
+
+		final PhenotypeDao dao = this.daoFactory.getPhenotypeDAO();
+		for (final ObservationDto observation : observations) {
+			final Phenotype phenotype = new Phenotype();
+			phenotype.setCreatedDate(new Date());
+			phenotype.setUpdatedDate(new Date());
+			phenotype.setObservableId(Integer.valueOf(observation.getObservationVariableDbId()));
+			phenotype.setValue(observation.getValue());
+			phenotype.setcValue(
+				this.resolveCategoricalIdValue(Integer.valueOf(observation.getObservationVariableDbId()), observation.getValue(),
+					validValuesForCategoricalVariables));
+			phenotype.setExperiment(experimentModelMap.get(observation.getObservationUnitDbId()));
+			this.setPhenotypeExternalReferences(observation, phenotype);
+			final Phenotype saved = dao.save(phenotype);
+			observationDbIds.add(saved.getPhenotypeId());
+		}
+		return observationDbIds;
+	}
+
+	private void setPhenotypeExternalReferences(final ObservationDto observationDto, final Phenotype phenotype) {
+		if (observationDto.getExternalReferences() != null) {
+			final List<PhenotypeExternalReference> references = new ArrayList<>();
+			observationDto.getExternalReferences().forEach(reference -> {
+				final PhenotypeExternalReference externalReference =
+						new PhenotypeExternalReference(phenotype, reference.getReferenceID(), reference.getReferenceSource());
+				references.add(externalReference);
+			});
+			phenotype.setExternalReferences(references);
+		}
+	}
+
+	private Map<Integer, List<ValueReference>> getCategoriesForCategoricalVariables(final List<ObservationDto> observations) {
+		final Set<Integer> variableIds =
+			observations.stream().map((o) -> Integer.valueOf(o.getObservationVariableDbId())).collect(Collectors.toSet());
+		return this.daoFactory.getCvTermRelationshipDao().getCategoriesForCategoricalVariables(new ArrayList<>(variableIds));
+
+	}
+
+	private Integer resolveCategoricalIdValue(final Integer variableId, final String value,
```

Comment:
To avoid returning null, better implement as
``` Optional<Integer> resolveCategoricalIdValue(final Integer variableId, final String value,
		final Map<Integer, List<ValueReference>> validValuesForCategoricalVariables)
```